velox/exec/tests/HashJoinTest.cpp

```diff
@@ -8112,4 +8112,85 @@ TEST_F(HashJoinTest, nanKeys) {
        makeFlatVector<int64_t>({1, 2})});
   facebook::velox::test::assertEqualVectors(expected, result);
 }
+
+TEST_F(HashJoinTest, combineSmallVectorsAfterFilter) {
+  // Verify low selectivity / small vectors are combined to 1 vector.
+  auto probeVectors = makeBatches(1, [&](auto /*unused*/) {
+    return makeRowVector(
+        {"t0", "t1"},
+        {
+            makeFlatVector<int32_t>(1'000, [](auto row) { return row; }),
+            makeFlatVector<int64_t>(1'000, [](auto row) { return row * 10; }),
+        });
+  });
+
+  auto buildVectors = makeBatches(3, [&](auto /*unused*/) {
+    return makeRowVector(
+        {"u0", "u1"},
+        {
+            makeFlatVector<int32_t>(
+                1'000, [](auto row) { return -100 + (row / 5); }),
+            makeFlatVector<int64_t>(
+                1'000, [](auto row) { return -1000 + (row / 5) * 10; }),
+        });
+  });
+
+  std::shared_ptr<TempFilePath> probeFile = TempFilePath::create();
+  writeToFile(probeFile->getPath(), probeVectors);
+
+  std::shared_ptr<TempFilePath> buildFile = TempFilePath::create();
+  writeToFile(buildFile->getPath(), buildVectors);
+
+  createDuckDbTable("t", probeVectors);
+  createDuckDbTable("u", buildVectors);
+
+  core::PlanNodeId probeScanId;
+  core::PlanNodeId buildScanId;
+  auto planNodeIdGenerator = std::make_shared<core::PlanNodeIdGenerator>();
+  auto makePlan = [&](bool nullAware) {
+    return PlanBuilder(planNodeIdGenerator)
+        .tableScan(asRowType(probeVectors[0]->type()))
+        .capturePlanNodeId(probeScanId)
+        .hashJoin(
+            {"t0"},
+            {"u0"},
+            PlanBuilder(planNodeIdGenerator)
+                .tableScan(asRowType(buildVectors[0]->type()))
+                .capturePlanNodeId(buildScanId)
+                .planNode(),
+            "(t1 + u1) % 3 = 0",
+            {"t0", "t1", "match"},
+            core::JoinType::kLeftSemiProject,
```

Comment:
Why kLeftSemiProject ? Should we tests all join types?

Reply:
Sure, will add tests for all join types.

Reply:
Updated, thanks.